### PR TITLE
46972299 - Improved 3D Unet for Segmenting HipMRI 3D Prostate MRI Dataset

### PR DESCRIPTION
**Pull request by art1f1c3R**
Task 4 - Hard Difficulty. This project uses the Improved 3D Unet deep learning model to segment and classify components of the HipMRI 3D Prostate MRI scans, considering 6 classes: background, bone, bladder, rectum, and prostate. This pull request contains the relevant files used to train, validate, and test the model, with a report description on usage, architecture explanation, and outcomes.
All files added are inside a folder `46972299_ImprovedUnet` inside the `recognition` folder. Files inside include:
- `dataset.py` includes the data loader and dataset objects that automatically load the images and masks and apply relevant transformations and data splits.
- `modules.py` contains the classes and actual architecture of the improved 3D Unet model.
- `metrics.py` includes classes for calculating loss metrics and accuracy.
- `train.py` is a script for running the training process and saving the model results.
- `predict.py` is a script for testing a trained (or even unfinished) model to analyse its performance and visualisate its results and metrics.
- `utils.py` contains provided code from the task and miscellaneous plotting and saving functions.
- `report_files` folder containing images used in the report
- `README.md` report as explained above.

### CODE REVIEW

**Comment by celinsze** [comment]
**This is an initial inspection, no action is required at this point**
Recognition problem: total 8 Marks
Problem solving: (1)
•	Segmentation results are outputted but with poor segmentation performance.
•	Average Dice loss was above 0.5 for all classes, representing a maximum Dice score of 0.5 for all labels.
Algorithm implementation: (3)
•	Algorithm implemented in PyTorch for the improved UNet3D model and functions.
Design: Good coding design (1)
Commenting: Sufficient comments throughout the code (1)
Difficulty:  (2)
•	3D Improved UNet – Hard
•	Problem attempted but not solved

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Segmentation results are outputted but with poor segmentation performance.</td><td>-3</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. UNet3D (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-3</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by hanemma7moud** [comment]
Carl Flottmann

**Comment by shakes76** [comment]
Segmentations not one hot encoded -2 additionally, results quality low -1
